### PR TITLE
April/patectgan

### PR DESCRIPTION
**Pull request by AprilXiaoyanLiu**
thought I created pull request but looks like I just committed to this branch. 
Some updates:
* implement ctgan with Pate Framework
* currently using cross-entropy loss

### CODE REVIEW

**Review by lurosenb — APPROVED** [review]
Verifying that the build passes, then will merge it over...

File: sdk/opendp/whitenoise/synthesizers/pytorch/nn/patectgan.py [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+from ctgan.conditional import ConditionalGenerator 
+
+from ctgan.models import Generator
+
+from ctgan.sampler import Sampler 
+
+from ctgan import CTGANSynthesizer
+
+import torchdp
+from torchdp import autograd_grad_sample
```
Comment (eedeleon) [inline sdk/opendp/whitenoise/synthesizers/pytorch/nn/patectgan.py:37]: neither of these imports are needed. patectgan does not use opacus right?

**Comment by eedeleon** [comment]
Pr looks good minus the torchdp imports

File: sdk/opendp/whitenoise/synthesizers/pytorch/nn/patectgan.py [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+from ctgan.transformer import DataTransformer
+from ctgan.conditional import ConditionalGenerator 
+
+from ctgan.models import Generator
+
+from ctgan.sampler import Sampler 
+
+from ctgan import CTGANSynthesizer
+
+import torchdp
```
Comment (eedeleon) [inline sdk/opendp/whitenoise/synthesizers/pytorch/nn/patectgan.py:35]: also not used. 

Not sure why the CI is not running. If you can remove this import, run the tests locally and send me a screen shot I can verify the tests passed. 

I am looking into why the build is currently not firing

File: sdk/opendp/whitenoise/synthesizers/pytorch/nn/patectgan.py [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+
+            fake = self.generator(fakez)
+            fakeact = self._apply_activate(fake)
+            data.append(fakeact.detach().cpu().numpy())
+        
+        data = np.concatenate(data, axis=0)
+        data = data[:n]
+
+        return self.transformer.inverse_transform(data, None)
+
```
Comment (eedeleon) [inline sdk/opendp/whitenoise/synthesizers/pytorch/nn/patectgan.py:373]: please check for new lines. There should be at most 1 or 2 new lines for spacing. More than that should be reduced

File: tests/sdk/synthesizers/test_pytorch_synthesizer.py [hunk 4]
```diff
@@ -53,4 +53,18 @@ def test_sample(self):
         self.dpctgan.fit(df, categorical_columns=['sex','educ','race','married'])
         sample_size = len(df)
         synth_data = self.dpctgan.sample(sample_size)
+        assert synth_data.shape == df.shape
+
+class TestPytorchDPSynthesizer_PATECTGAN:
+    def setup(self):
+        self.patectgan = PytorchDPSynthesizer(PATECTGAN(), None)
```
Comment (eedeleon) [inline tests/sdk/synthesizers/test_pytorch_synthesizer.py:60]: Given the change in name to opacus, this is a reasonable name now :) Which is niced. The is Pytorch DP  with PATE. No package name conflict anymore after we update to opacus

**Review by eedeleon — APPROVED** [review]
Looks good after that minor change

File: sdk/opendp/whitenoise/synthesizers/pytorch/nn/patectgan.py (hunk 2 above)
Comment (lurosenb) [reply sdk/opendp/whitenoise/synthesizers/pytorch/nn/patectgan.py]: new_tests_patectgan

Comment (lurosenb) [reply sdk/opendp/whitenoise/synthesizers/pytorch/nn/patectgan.py]: just ran the 2 new tests that were added locally, should be good to merge

**Comment by eedeleon** [comment]
Squashing although the build did not run. Looking into a longer term fix for this. @lurosenb  tagged the new builds and we are queueing master shortly
